### PR TITLE
[ticket/15274] Allow "custom" migrations to use parameters

### PR DESCRIPTION
**Pull request by javiexin**
Add the possibility to have custom migrations with parameters,
allowing the use of a single function for several uses.

PHPBB3-15274

Checklist:

- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB3-15274

### CODE REVIEW

**Comment by 3D-I** [comment]
To adhere to the documentation.

File: tests/dbal/migration/if_params.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	{
+		return array();
+	}
+
+	function update_data()
+	{
+		return array(
+			array('if', array(
+				true,
+				array('custom', array(array(&$this, 'test'), array('true'))),
```
Comment (Nicofuma) [inline tests/dbal/migration/if_params.php]: remove the reference

File: tests/dbal/migration/recall_params.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+{
+	function update_schema()
+	{
+		return array();
+	}
+
+	function update_data()
+	{
+		return array(
+			array('custom', array(array(&$this, 'test_call'), array(5))),
```
Comment (Nicofuma) [inline tests/dbal/migration/recall_params.php]: removes the reference

File: tests/dbal/migration/if_params.php (hunk 1 above)
Comment (javiexin) [inline tests/dbal/migration/if_params.php]: Sure.  I will also remove it in the original tests that I copied to create these.

Please, update the documentation as well (tutorial): https://area51.phpbb.com/docs/dev/32x/migrations/data_changes.html#example

File: tests/dbal/migration/recall_params.php (hunk 2 above)
Comment (javiexin) [inline tests/dbal/migration/recall_params.php]: Same here, removed from here and from original test copied to create this one.

File: tests/dbal/migration/if_params.php (hunk 1 above)
Comment (Nicofuma) [inline tests/dbal/migration/if_params.php]: you can do it yourself https://github.com/phpbb/documentation/blob/3.1.x/development/migrations/data_changes.rst

Comment (javiexin) [inline tests/dbal/migration/if_params.php]: Open an issue there, or is there a separate tracker?  Then, should I fork, edit, and submit a PR against that one?  Only for fixes (remove reference, clarify that no params are allowed), or also to document this change (params allowed, maybe with example)?

Comment (Nicofuma) [inline tests/dbal/migration/if_params.php]: There is no specific tracker for the documentation. We use the github issues when needed.
Here, you can send a PR directly with the doc fixes and the new doc for what you have added

Comment (Nicofuma) [inline tests/dbal/migration/if_params.php]: Not that the documentation use the same branches than phpbb/phpbb (3.1.x, 3.2.x and master)

Comment (javiexin) [inline tests/dbal/migration/if_params.php]: Sorry, but "Permission to phpbb/documentation.git denied to javiexin."

**Comment by Nicofuma** [comment]
@marc1706 3.1 or 3.2?

**Comment by Nicofuma** [comment]
By the way, there is one limitation that needs to be documented imo: it is not compatible with multi-step processes (https://area51.phpbb.com/docs/dev/32x/migrations/data_changes.html#multi-step-processes)

Otherwise, @javiexin are you planning to update the documentation?

**Comment by javiexin** [comment]
Why not? It is fully compatible with multi-step, as you can see from one of the already existing tests.

As for updating documentation, I tried, but was denied access.

**Comment by Nicofuma** [comment]
what I mean is that you can't do both at the same time

What do you mean by "was denied access"?

File: tests/dbal/migration/if_params.php (hunk 1 above)
Comment (Nicofuma) [inline tests/dbal/migration/if_params.php]: when did you get that? it is a public repo

**Comment by javiexin** [comment]
https://github.com/javiexin/phpbb/blob/55d5ed79fb407dcfc30449df3885500996c77baf/tests/dbal/migration/recall_params.php

File: tests/dbal/migration/if_params.php (hunk 1 above)
Comment (javiexin) [inline tests/dbal/migration/if_params.php]: Exactly when I wrote that...

**Comment by Nicofuma** [comment]
nice, I missed it

**Comment by javiexin** [comment]
@Nicofuma I understand that this is NOT going to be merged into 3.1, so if I update documentation, it should be against 3.2 branch, correct?  Should I indicate that this will be a 3.2.2 new feature?

**Comment by Nicofuma** [comment]
It should be against 3.2 yes.
For the 3.2.2 indication I don't know if we have a such thing anywhere in the doc. But it is a good idea

**Comment by javiexin** [comment]
Ok, will try to update the docs again.  If I have problems, I will advise again.

**Comment by javiexin** [comment]
@Nicofuma Again, same error when pushing to documentation repo:

> remote: Permission to phpbb/documentation.git denied to javiexin.
> fatal: unable to access 'https://github.com/phpbb/documentation.git/': The requested URL returned error: 403
> Pushing to https://github.com/phpbb/documentation.git

This is to the 3.1 branch, to fix the unneeded references (&) in the documentation.  I have not tested in 3.2 branch though.

**Comment by Nicofuma** [comment]
off course, push to your fork and open a PR, do not push to phpbb/documentation

**Comment by javiexin** [comment]
Of course!!  Am I stupid or what!  Sorry to bother you with this...  Fix for error in documentation sent as PR against 3.1, now new functionality will be sent against 3.2 branch.

**Comment by Nicofuma** [comment]
thanks

**Comment by javiexin** [comment]
So finally 3.3?  Oh, well...

**Pull request merged**